src/webgpu/api/operation/sampling/sampler_texture.spec.ts

```diff
@@ -0,0 +1,270 @@
+        `    @group(${groups.length - 1}) @binding(${binding}) var ${resourceId}: ${resourceType};`
+      );
+      bindGroupEntries.push({
+        binding,
+        resource,
+      });
+      bindGroupLayoutEntries.push({
+        binding,
+        visibility: stage === 0 ? GPUShaderStage.VERTEX : GPUShaderStage.FRAGMENT,
+        ...(resource instanceof GPUSampler
+          ? {
+              sampler: {},
+            }
+          : {
+              texture: {},
+            }),
+      });
+    }
+
+    function addTexture(stage: number, textureNum: number) {
+      const textureId = `tex${stage}_${textureNum}`;
+      if (!textureIds.has(textureId)) {
+        textureIds.add(textureId);
+        const texture = t.createTextureTracked({
+          format: 'rgba8unorm',
+          size: [maxSamplersPerShaderStage, 1],
+          usage: GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_DST,
+        });
+        textures.push(texture);
+        // Encode an rgba8unorm texture with rg16uint data where each texel is
+        // [(textureId + 1) | (stage << 15), {samplerId + 1}]
+        // The +1 is to avoid 0.
+        const data = new Uint16Array(maxSamplersPerShaderStage * 2);
+        const rg = (textureNum + 1) | (stage << 15);
+        for (let x = 0; x < maxSamplersPerShaderStage; ++x) {
+          const offset = x * 2;
+          const samplerNum = x + 1;
+          data[offset + 0] = rg;
+          data[offset + 1] = samplerNum;
+        }
+        device.queue.writeTexture({ texture }, data, {}, [maxSamplersPerShaderStage]);
+        addResource(stage, textureId, texture.createView());
+      }
+      return textureId;
+    }
+
+    const kAddressModes = ['repeat', 'clamp-to-edge', 'mirror-repeat'] as const;
+    const getAddressMode = (hash: number, depth: number) => {
+      return kAddressModes[
+        ((hash / Math.pow(kAddressModes.length, depth)) | 0) % kAddressModes.length
+      ];
+    };
+
+    function addSampler(stage: number, samplerNum: number) {
+      const samplerId = `smp${stage}_${samplerNum}`;
+      if (!samplerIds.has(samplerId)) {
+        const samplerNum = samplerIds.size;
+        samplerIds.add(samplerId);
+        // try to make each sampler unique. This is because some backends
+        // coalesce samplers with the same settings.
+        const addressHash = samplerNum >> 3;
+        const sampler = device.createSampler({
+          minFilter: samplerNum & 1 ? 'linear' : 'nearest',
+          magFilter: samplerNum & 2 ? 'linear' : 'nearest',
+          mipmapFilter: samplerNum & 4 ? 'linear' : 'nearest',
+          addressModeU: getAddressMode(addressHash, 0),
+          addressModeV: getAddressMode(addressHash, 1),
+          addressModeW: getAddressMode(addressHash, 2),
+        });
+        addResource(stage, samplerId, sampler);
+      }
+      return samplerId;
+    }
+
+    // Note: We are storing textureId, samplerId in th texture. That suggests we could use rgba32uint
```

Comment:
nit: `in th texture` -> `in the texture`?

Reply:
```suggestion
    // Note: We are storing textureId, samplerId in the texture. That suggests we could use rgba32uint
```